new-chatbot-app/client/src/App.jsx

```diff
@@ -15,156 +10,85 @@ import {
-  const [attemptedConnection, setAttemptedConnection] = useState(false);
+  const firstSession = useRef(true);
+  const { socket, isConnected, attemptedConnection, setIsConnected, setAttemptedConnection } = useContext(SocketContext);
+  const { setMessage, message, addMessage, setIsTyping, } = useContext(MessageContext);
 
-  /**
-   * Hook to connect to the socket and listen for messages
-   */
   useEffect(() => {
-    // Will return empty array if there are no messages in the session storage
-    const storedMessages =
-      JSON.parse(sessionStorage.getItem('chat_messages')) || [];
-    setMessages(storedMessages);
-
-    // Set up URL from environment variables
-    const url = `${retrieveEnvironmentVariable('VITE_BACKEND_URL')}:${retrieveEnvironmentVariable(
-      'VITE_BACKEND_PORT',
-    )}`;
-    // Connect to the socket server with option to use websocket and disable upgrade
-    const socketIo = io(url, { transports: ['websocket'], upgrade: false });
-
-    socketIo.on('connect', () => {
-      setIsConnected(true); // Update the state to indicate that the connection is established
-      setAttemptedConnection(true); // Update the state to indicate that the connection has been attempted
-      if (!welcomeMessageShown) {
-        // Send the welcome message only once per session
+    socket.on('connect', () => {
+      if (firstSession.current) {
         const welcomeMessage = {
           text: 'Hi this is the Library Smart Chatbot. How may I help you?',
           sender: 'chatbot',
         };
-        setMessages((prevMessages) => {
+        setMessage((prevMessages) => {
           const updatedMessages = [...prevMessages, welcomeMessage];
-          // Store the messages in the session storage
           sessionStorage.setItem(
             'chat_messages',
             JSON.stringify(updatedMessages),
           );
           return updatedMessages;
         });
-        setWelcomeMessageShown(true); // Update the state to ensure the welcome message is shown only once
+        firstSession.current = false;
       }
+      setIsConnected(true);
+      setAttemptedConnection(true);
       setIsTyping(false);
     });
 
-    // Listen from the server
-    socketIo.on('message', function (message) {
+    socket.on('message', function (message) {
       setIsTyping(false);
       addMessage(message, 'chatbot');
     });
 
-    socketIo.on('disconnect', function () {
+    socket.on('disconnect', function () {
       setIsConnected(false);
       setAttemptedConnection(true);
     });
 
-    socketIo.on('connect_error', (error) => {
+    socket.on('connect_error', (error) => {
       console.error('Connection Error:', error);
       setIsConnected(false);
       setAttemptedConnection(true);
     });
 
-    socketIo.on('connect_timeout', (timeout) => {
+    socket.on('connect_timeout', (timeout) => {
       console.error('Connection Timeout:', timeout);
       setIsConnected(false);
       setAttemptedConnection(true);
     });
 
```

Comment:
I think these socket.on should be in the SocketContextProvider, then SocketContextProvider would export the variables to notify other component if there is new incoming message or client disconnects.

Reply:
I moved the whole useEffect to SocketContextProvider and it seems to be working